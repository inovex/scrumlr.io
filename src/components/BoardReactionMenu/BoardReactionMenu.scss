@import "src/constants/style";

$container-height: 64px;
$reaction-size: 54px;

.board-reactions__root {
  position: absolute;
  display: flex;
  justify-content: center;
  width: 100%;
  bottom: $column__border-width;
}

.board-reactions__container {
  height: $container-height;

  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  gap: 8px;

  background-color: $color-white-one;
  border-radius: $rounded--large;
  box-shadow: $box-shadow--light;

  padding: 0 $padding--default;

  user-select: none;

  z-index: $board-reaction-z-index;
}

.board-reactions__item {
  display: flex;
  justify-content: center;
  align-items: center;

  cursor: pointer;
}

.board-reactions__reaction {
  all: unset; // reset button
  cursor: pointer;

  width: $reaction-size;
  height: $reaction-size;

  font-size: $text--xl;

  border-radius: $rounded--full;

  display: flex;
  justify-content: center;
  align-items: center;

  &:hover {
    background-color: $hover-background-light;
  }

  &:focus-visible {
    // lose focus after click
    background-color: $color-icon-light-blue;
  }

  &--disabled {
    filter: grayscale(100%);
    cursor: not-allowed;
  }
}

.board-reactions__close {
  all: unset;
  width: $icon--large;
  height: $icon--large;
  color: $color-dark-mode-note;
  border-radius: $rounded--full;

  &:focus {
    background-color: $color-icon-light-blue;
  }
}

@media #{$smartphone} {
  .board-reactions__root {
    bottom: $column__border-width + $margin--medium + 32px;
  }

  .board-reactions__close {
    display: none;
  }
}

[theme="dark"] {
  .board-reactions__container {
    background-color: $color-dark-mode-note;
  }

  .board-reactions__reaction {
    &:hover {
      background-color: $hover-background-dark;
    }

    &:focus-visible {
      background-color: $tooltip-background-color--dark;
    }
  }

  .board-reactions__close {
    color: $color-white;
  }
}

@media screen and (max-width: 400px) {
  .board-reactions__container {
    padding: 0 $padding--small;
    gap: 2px;
  }
}
