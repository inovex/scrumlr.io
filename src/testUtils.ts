export type AssertTypeEqual<T, Expected> =
    T extends Expected
        ? (Expected extends T ? true : never)
        : never;