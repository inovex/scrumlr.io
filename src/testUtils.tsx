import {render, RenderOptions} from "@testing-library/react";
import {I18nextProvider} from "react-i18next";
import {FC, PropsWithChildren, ReactElement} from "react";
import {MemoryRouter} from "react-router";
import {Provider} from "react-redux";
import getTestStore from "utils/test/getTestStore";
import i18n from "./i18nTest";

export type AssertTypeEqual<T, Expected> = T extends Expected ? (Expected extends T ? true : never) : never;

const AllTheProvidersWithoutRouter: FC<PropsWithChildren> = ({children}) => (
  <I18nextProvider i18n={i18n}>
    <Provider store={getTestStore()}>{children}</Provider>
  </I18nextProvider>
);

const AllTheProviders: FC<PropsWithChildren> = ({children}) => (
  <MemoryRouter>
    <AllTheProvidersWithoutRouter>{children}</AllTheProvidersWithoutRouter>
  </MemoryRouter>
);

const customRender = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) => render(ui, {wrapper: AllTheProviders, ...options});
const customRenderWithoutRouter = (ui: ReactElement, options?: Omit<RenderOptions & {container: Element}, "wrapper">) =>
  render(ui, {wrapper: AllTheProvidersWithoutRouter, ...options});

export {customRender as render, customRenderWithoutRouter as renderWithoutRouter};
