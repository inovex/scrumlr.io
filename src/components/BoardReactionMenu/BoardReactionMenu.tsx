import React from "react";
import {ReactComponent as CloseIcon} from "assets/icon-close.svg";
import {ReactionType} from "types/reaction";
import {BOARD_REACTION_EMOJI_MAP} from "types/boardReaction";
import {Actions} from "store/action";
import {useDispatch} from "react-redux";
import {useHotkeys} from "react-hotkeys-hook";
import {useAppSelector} from "store";
import {Toast} from "utils/Toast";
import {useTranslation} from "react-i18next";
import classNames from "classnames";
import {useDelayedReset} from "utils/hooks/useDelayedReset";
import "./BoardReactionMenu.scss";

type BoardReactionMenuProps = {
  close: () => void;
};

const REACTION_DEBOUNCE_TIME = 300; // milliseconds

export const BoardReactionMenu = (props: BoardReactionMenuProps) => {
  const dispatch = useDispatch();
  const {t} = useTranslation();

  const boardReactions = [...BOARD_REACTION_EMOJI_MAP];

  const showBoardReactions = useAppSelector((state) => state.view.showBoardReactions);

  const [debounce, resetDebounce] = useDelayedReset<boolean>(false, true, REACTION_DEBOUNCE_TIME);

  const onClickReaction = (e: React.MouseEvent<HTMLButtonElement> | KeyboardEvent, reaction: ReactionType) => {
    e.stopPropagation();
    if (!showBoardReactions) {
      Toast.info({
        title: t("Toast.boardReactionsDisabled"),
        buttons: [t("Toast.enable")],
        firstButtonOnClick: () => dispatch(Actions.setShowBoardReactions(true)),
      });
    } else if (debounce) {
      // not ready yet
    } else {
      dispatch(Actions.addBoardReaction(reaction));
      resetDebounce();
    }
  };

  // hotkey is converted to a number which is then used as the index for the reaction type
  useHotkeys(["1", "2", "3", "4", "5"], (e, k) => onClickReaction(e, boardReactions[+k.keys![0] - 1][0]));

  return (
    <div className="board-reactions__root">
      <div className="board-reactions__container">
        {boardReactions.map(([reactionType, emoji], index) => (
          <button
            key={reactionType}
            className={classNames("board-reactions__item board-reactions__reaction", {"board-reactions__reaction--disabled": !showBoardReactions})}
            aria-disabled={!showBoardReactions}
            aria-label={t("BoardReactionsMenu.react", {reaction: reactionType, shortcut: index + 1})}
            title={t("BoardReactionsMenu.react", {reaction: reactionType, shortcut: index + 1})}
            onClick={(e) => onClickReaction(e, reactionType)}
          >
            <span>{emoji}</span>
          </button>
        ))}

        <button className="board-reactions__item board-reactions__close" onClick={props.close} tabIndex={0} aria-hidden>
          <CloseIcon />
        </button>
      </div>
    </div>
  );
};
