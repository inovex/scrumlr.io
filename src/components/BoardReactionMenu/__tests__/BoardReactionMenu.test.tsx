import {BoardReactionMenu} from "../BoardReactionMenu";
import {render} from "testUtils";
import {fireEvent} from "@testing-library/react";

describe("board reaction menu", () => {
  it("should render correctly", () => {
    const mockClose = jest.fn();
    const boardReactionMenu = render(<BoardReactionMenu showMenu={true} close={mockClose} />);
    expect(boardReactionMenu.container).toMatchSnapshot();
  });

  it("should close on clicking close button", () => {
    const mockClose = jest.fn();
    const boardReactionMenu = render(<BoardReactionMenu showMenu={true} close={mockClose} />);
    fireEvent.click(boardReactionMenu.container.getElementsByClassName("board-reactions-menu__close")[0]);
    expect(mockClose).toHaveBeenCalled();
  });
});
