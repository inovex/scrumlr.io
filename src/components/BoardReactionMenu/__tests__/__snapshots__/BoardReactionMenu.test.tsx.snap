// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`board reaction menu should render correctly 1`] = `
<div>
  <div
    class="board-reactions-menu"
    style="opacity: 0; transform: scale(0.3, 0.9) translateY(100%); pointer-events: none;"
  >
    <button
      aria-disabled="false"
      aria-label="React with tada [1]"
      class="board-reactions-menu__item board-reactions-menu__reaction"
      title="React with tada [1]"
    >
      <span>
        🎉
      </span>
    </button>
    <button
      aria-disabled="false"
      aria-label="React with applause [2]"
      class="board-reactions-menu__item board-reactions-menu__reaction"
      title="React with applause [2]"
    >
      <span>
        👏
      </span>
    </button>
    <button
      aria-disabled="false"
      aria-label="React with heart [3]"
      class="board-reactions-menu__item board-reactions-menu__reaction"
      title="React with heart [3]"
    >
      <span>
        💖
      </span>
    </button>
    <button
      aria-disabled="false"
      aria-label="React with like [4]"
      class="board-reactions-menu__item board-reactions-menu__reaction"
      title="React with like [4]"
    >
      <span>
        👍
      </span>
    </button>
    <button
      aria-disabled="false"
      aria-label="React with dislike [5]"
      class="board-reactions-menu__item board-reactions-menu__reaction"
      title="React with dislike [5]"
    >
      <span>
        👎
      </span>
    </button>
    <button
      aria-hidden="true"
      class="board-reactions-menu__item board-reactions-menu__close"
      tabindex="0"
    >
      <svg>
        icon-close.svg
      </svg>
    </button>
  </div>
</div>
`;
